src/Bicep.Core/PrettyPrint/DocumentBuildVisitor.cs

```diff
@@ -195,32 +191,56 @@ public ILinkedDocument BuildDocument(SyntaxBase syntax)
                  return Spread(Concat(openParen, itemVariable, comma), Concat(indexVariable, closeParen));
              });
 
-        public override void VisitFunctionCallSyntax(FunctionCallSyntax syntax) =>
-            this.Build(() => base.VisitFunctionCallSyntax(syntax), children =>
-            {
-                Debug.Assert(children.Length >= 3);
+        private void VisitCommaAndNewLineSeparated(ImmutableArray<SyntaxBase> nodes)
+        {
+            var hasTrailingNewline = nodes.Length > 0 && nodes[^1] is Token { Type: TokenType.NewLine };
+            this.Build(() => {
+                for (var i = 0; i < nodes.Length; i++)
+                {
+                    this.Visit(nodes[i]);
+
+                    if (i < nodes.Length - 1 &&
+                        nodes[i] is Token { Type: TokenType.Comma } &&
+                        nodes[i + 1] is not Token { Type: TokenType.NewLine })
+                    {
+                        this.PushDocument(Space);
+                    }
+                }
+            }, children => {
+                // This logic ensures that syntax with only a single child is not doubly-nested,
+                // and that the final newline does not cause the next piece of text to be indented
+                // e.g. 'bar' in the following is only indented once, and '})' is not indented:
+                //   foo({
+                //     bar: 123
+                //   })
+                var nestedChildren = Concat(hasTrailingNewline ? children[..^1] : children);
+                var newLine = hasTrailingNewline ? children[^1] : Nil;
+
+                if (children.Length > 1)
+                {
+                    nestedChildren = nestedChildren.Nest();
+                }
 
-                ILinkedDocument name = children[0];
-                ILinkedDocument openParen = children[1];
-                ILinkedDocument arguments = Spread(children.Skip(2).SkipLast(1));
-                ILinkedDocument closeParen = children[^1];
+                return Concat(nestedChildren, newLine);
```

Comment:
It would be great to add traling whitespaces for objects and arrays when there's no newlines before the first item and after the last item. For examples:
```
var foo = [ 1, 2, 3, 4
]

var bar = [
  1, 2, 3, 4 ]

var baz = [ 1, 2, 3, 4 ]
```

Reply:
Good point! I've fixed this.